includes/class-wc-payments-payment-request-button-handler.php

```diff
@@ -224,7 +225,16 @@ public function get_product_price( $product, ?bool $is_deposit = null ) {
 				$is_deposit = 'deposit' === WC_Deposits_Product_Manager::get_deposit_selected_type( $product->get_id() );
 			}
 			if ( $is_deposit ) {
-				$base_price = WC_Deposits_Product_Manager::get_deposit_amount( $product, 0, 'display', $base_price );
+				$deposit_type = WC_Deposits_Product_Manager::get_deposit_type( $product->get_id() );
+				if ( 'plan' === $deposit_type && 0 === $deposit_plan_id ) {
+					// Default to first (default) plan if no plan is specified.
+					$available_plan_ids = WC_Deposits_Plans_Manager::get_plan_ids_for_product( $product->get_id() );
+					if ( ! empty( $available_plan_ids ) ) {
+						$deposit_plan_id = $available_plan_ids[0];
+					}
+				}
+
+				$base_price = WC_Deposits_Product_Manager::get_deposit_amount( $product, $deposit_plan_id, 'display', $base_price );
```

Comment:
Unfortunately not. While I agree this logic seems to be more related to the Deposits extensions, when we use a payment request button, we skip key parts of the process, so we have to recreate it, [like we do for variations](https://github.com/Automattic/woocommerce-payments/blob/7448731cb746d6e6832d389557ba1fad610f03e0/includes/class-wc-payments-payment-request-button-handler.php#L257-L276).
For example, the method `get_product_data` is used to get the payment's initial data, which will be rendered as JSON with the page load. So, the product can't be added to the cart, but we still need to get the exact same data as if it were. This is something that is only ever done for payment requests, so this logic belongs here. Also notice in this method we call `get_product_price` without any arguments, which mean it will use the default data, so we could let WC_Deposits handle everything and return the default amount, but it is better to do it manually so it can be reused by the second method, `ajax_get_selected_product_data`.
When we call this method, we are creating a new cart manually, not using the "Add to cart" button, so we need to handle all the data we can by ourselves. Because of that, we can't have `get_product_price` always return the default amount for the deposit; we need to handle the selected options. So, the code that returns the price needs to support provided values instead of always using the default.

> Is Default to first (default) plan if no plan is specified something that should be done by Deposits?

Well, it is, but how is it specified if not by us? Updating 
`WC_Deposits::get_deposit_amount()` to read from `$_POST` is a bad pattern. Request params should be read in the controller and then we pass the data to `get_deposit_amount`.
We are passing these variables because we want to be able to overwrite `$is_deposit` and `$deposit_plan_id` as needed.

Reply:
Thanks for your response, Guilherme! Below, I'm raising a few more thoughts and open questions; sorry for prolonging this topic, I do appreciate your knowledge in this area and find it really helpful to learn, thanks to those discussions. I don't want this thread to block the whole PR, so please feel free to merge the PR, and, in the worst case, we can improve this small area separately if there is such a need. 

> For example, the method get_product_data is used to get the payment's initial data, which will be rendered as JSON with the page load. So, the product can't be added to the cart, but we still need to get the exact same data as if it were. This is something that is only ever done for payment requests, so this logic belongs here

As I understand from your explanation, there are two steps to building a dataset for a PRB, and the one described here is the first step, where we provide initial data, which might (and most probably will) change after running the second step through `wc_ajax_wcpay_get_selected_product_data` hook. Given that it's just initial data, which is subject to change, can we prepare it as if no deposit applies to a product? During the second dataset update, we will provide the real deposit data for a product. If such an approach would make sense, then we could 

1. Exclude deposit-related logic which updates the `$base_price`, from `get_product_price()`
2. Add deposit-related logic within `get_product_data()` right after `get_product_price()`, and make sure this deposit-related logic runs only during the second step (`ajax_get_selected_product_data`) and is skipped on the first step. We could distinguish between steps with `wc_deposit_option/wc_deposit_payment_plan` request params I guess.

This way, we will decouple a PRB-specific logic, which manages the flow for when getting the deposit-related price, from the actual get-deposit-related-price logic. Now, having this, we would know that we only calculate deposits when needed. And now, coming back to how it could make it, I still think (maybe I'm missing something) that Deposits could encapsulate all the steps within lines `228-237` into its own single method. 

> the code that returns the price needs to support provided values instead of always using the default

Yes, and I think the main point of my questions here is exactly if this support of provided deposit values should be within the Deposits area of responsibility. 


> Updating
WC_Deposits::get_deposit_amount() to read from $_POST is a bad pattern. Request params should be read in the controller and then we pass the data to get_deposit_amount.

I completely agree and thank you for bringing this up. Sorry that I missed this obvious thing. So we still are forwarding `$deposit_plan_id` further. Then the single method from Deposits could get the plan id as an argument.

Reply:
> the first step, where we provide initial data, which might (and most probably will) change after running the second step through

The second step never runs for virtual products. Only products that require shipping update after initialized.

> Given that it's just initial data, which is subject to change, can we prepare it as if no deposit applies to a product?

Because of the issue above, if we start it with the incorrect value it will never be updated to the correct value if you're paying for a virtual product. The problem for physical products would be the price flashing wrong for a few seconds.

> Yes, and I think the main point of my questions here is exactly if this support of provided deposit values should be within the Deposits area of responsibility.

I agree it should be that plugin responsibility to set the price, but here we need the price in different circumstances, so it makes sense for me we have to ask for it. In an ideal scenario we would have the item properly added to the cart (not by us), then let the cart do the calculations and use the result. The problem is that in both of the methods we have here, we can't do that, this makes this logic specific to wcpay. So we need to either have wcpay with code to add compatibility to Deposits or the opposite. But I think it is something too specific for Apple Pay and will feel even more out of place if done in WC Deposits. What do you think?